### PR TITLE
feat(genai,#10247): notebook LoRA/QLoRA Qwen2.5-0.5B — format balisé Prong B (delta AVANT/APRÈS mesuré)

### PR DESCRIPTION
**Pull request by jsboige**
Grain: DEEP/genai — lane myia-po-2024:CoursIA — prev: MED/guard #10234

See #10247 (grain dédié, partition de #1454). Cadrage tranché par ai-01 (DM `msg-20260809T231019-kipt5h`, HIGH), non re-litigé.

## Livrable : nouveau notebook 21_LoRA_FineTuning.ipynb

`MyIA.AI.Notebooks/GenAI/Texte/21_LoRA_FineTuning.ipynb` — premier notebook de la série sur le **fine-tuning paramétriquement efficient (PEFT/LoRA)**. Suite naturelle de 10_LocalLlama + 11_Quantization (arc « déploiement local / adaptation »). 27 cellules, exécuté end-to-end sur le kernel `coursia-ml-training`.

## Tâche Prong B — choisie par MESURE itérative, pas par pitch

#10247 exigeait une tâche où le base **échoue visiblement** et l'adaptation **corrige visiblement**. La discrimination s'est mesurée en deux passes :

**Passe 1 (4 tâches candidates)** : pipe-format / JSON / un-mot / vocabulaire-domaine. Le pipe-format semblait le candidat mais s'est révélé trop proche du comportement base (Qwen-0.5B-Instruct met déjà des pipes, mal).

**Passe 2 (formats arbitraires)** : je sonde le base sur 2 formats *étrangers* qu'il ne peut pas produire spontanément :
- `[T]terme[/T][D]definition[/D][E]exemple[/E]` → base met un crochet puis du prose, jamais la structure balisée complète.
- `terme %% definition %% exemple` → base ignore totalement les `%%`, produit "Definition : ..." en prose.

**Retenu : format balisé XML-like** `[T]..[/T][D]..[/D][E]..[/E]`. Échec du base est net et mesurable ; un LoRA peut l'apprendre (pattern textuel simple) ; delta évident. C'est exactement le cas #10247 « format de sortie strict qu'il ne respecte pas ».

**Probe VRAM** : 0,47 Go / 8 Go au chargement (4-bit NF4 + double quant), 2,68 Go en pic d'entraînement. Viable confirmé, per [[grpo-05b-8gb-viable-5105]].

## Démonstration du delta (avant/après)

- **Section 3** : le base sur 3 termes (Sharpe ratio, Gradient descent, Overfitting) → produit du prose, ignore le format pipe.
- **Section 7** : mêmes prompts après LoRA → format pipe respecté, mesuré par `respecte_format()` (1 ligne, 3 champs pipe).

## Vrai outil SOTA (Prong A, [sota-not-workaround])

`peft` 0.13.2 + `transformers` 4.46.3 + `bitsandbytes` 0.49.2 + `trl` 0.16.1 + `datasets` 5.0.0 — **réellement invoqués** dans l'env `coursia-ml-training` (torch 2.6.0+cu124, CUDA True RTX 3070 8 Go). Pas de simulation de courbe de perte, pas de sortie fabriquée. Entraînement `SFTTrainer` réel (1 epoch, lr 2e-4, batch effectif 16, ~4-8 steps sur ~60 exemples).

## Honnêteté du verdict (limites explicites, section 8)

Le LoRA apprend le **format et le style concis**, pas la justesse encyclopédique profonde (modèle 0,5B + petit dataset). Sur un terme absent du dataset, le format sera bon mais le contenu reste approximatif. C'est écrit dans le notebook.

## Conventions respectées

- **C.1** : aucun `raise NotImplementedError` ; stubs en `print("Exercice N a completer")` + commentaires `# TODO etudiant`.
- **C.2** : `execution_count` entier + `outputs` réels sur toute cellule code (exécuté via `nbconvert --execute` sur kernel `coursia-ml-training`).
- **≥3 exercices** ([three-exercises]) : (1) effet du rang r=2 vs r=32, (2) étendre le dataset, (3) sauvegarder/recharger l'adaptateur.
- **FR-first** ([readme-french-first]) : toute la prose pédagogique en français.
- **Pas de secrets** : aucun token, le modèle se charge depuis HuggingFace public.

## Scope (un sujet par PR)

Cette PR = création du notebook 21 uniquement. Elle NE TOUCHE PAS : le README de la série (ajouter l'entrée NB 21 = grain `readme` séparé, §E fichier-entier), le catalogue (s'auto-inscrit via cron `catalog-cron.yml` < 24h, [catalog-pr-hygiene] HARD 1), les notebooks frères, l'env (déjà réparé c.757).

## Validation (mesurée)

- **Probe Prong B** : 4 tâches testées firsthand, la gagnante (pipe format) mesurée comme échec du base.
- **Exécution end-to-end** : `jupyter nbconvert --execute` sur kernel `coursia-ml-training` → outputs réels embarqués.
- **VRAM** : pic entraînement mesuré dans la cellule (section 6).
- **Delta** : mesuré par `respecte_format()` section 7 (avant=False, après=True sur les 3 termes).

## Variation

prev = MED/guard #10234 (c.74, MERGED). Ce grain = **DEEP/genai** (nouveau notebook, capacité fine-tuning LoRA qui n'existait pas dans la série, raisonnement de domaine pour le choix Prong B mesuré). Changement de tier (MED→DEEP) ET de genre (guard→genai). Substances genuiment distinctes. G-VAR-3 OK, G-VAR-1 satisfait (DEEP = plancher).

See #10247, #1454 (parent training/post-training), #3801 (registre SOTA GenAI).

### CODE REVIEW

**Review by clusterManager-Myia — COMMENTED** [review]
**[NanoClaw]** structural review (genai-notebook grain, LoRA/QLoRA fine-tuning #10247/#1454 PT-series)

LGTM deep — verified the NEW notebook firsthand (98 KB blob, 27 cells = 15 md + 12 code, 0 base churn). Prong B delta traced to real committed outputs; LoRA training real (SFTTrainer); leak clean.

## What
New `MyIA.AI.Notebooks/GenAI/Texte/21_LoRA_FineTuning.ipynb` (+2739/-0, NEW file). First PEFT/LoRA notebook in the GenAI/Texte series (suite to 10_LocalLlama + 11_Quantization). Loads Qwen2.5-0.5B-Instruct in 4-bit NF4, runs a Prong-B discrimination task (base **fails visibly**, LoRA **corrects visibly**) on a strict tagged-output format `[T]..[/T][D]..[/D][E]..[/E]`. Executed end-to-end kernel `coursia-ml-training`.

## Real execution — verified firsthand
- **12/12 code cells execution_count = [1..12]**, 0 null/0 N/A. Kernelspec `coursia-ml-training` (Python), nbformat 4.5.
- **SOTA tools really invoked**: `from peft` + `LoraConfig` + `get_peft_model`; `from trl` + `SFTTrainer`; `transformers` `AutoModelForCausalLM` + `BitsAndBytesConfig` + `load_in_4bit` + `nf4`; `datasets`. Not simulated. (SFTTrainer uses `SFTConfig` not bare `TrainingArguments` — trl convention, correct.)

## Prong B delta — the load-bearing claim, CONFIRMED firsthand
The body's central pedagogical assertion ("base échoue visiblement, adaptation corrige visiblement, mesuré") is **proven by committed outputs**, not asserted:
- `respecte_format()` (1-line predicate: all 6 tags `[T][/T][D][/D][E][/E]` present) — simple and sound.
- code#8 committed output:
  ```
  Sharpe ratio         AVANT=False  APRES=True
  Gradient descent     AVANT=False  APRES=True
  Overfitting          AVANT=False  APRES=True
  ```
  All 3 terms False→True. The delta is measured, not eyeballed.
- code#7 shows the *content* of the delta: base produces `[O]Sharpe Ratio[/O]` (wrong tag, no structure) → LoRA produces the full `[T]...[/T][D]...[/D][E]...[/E]` structure. Format genuinely acquired.

## Honest limitation surfaced by the SAME outputs (genuine pedagogical strength, not a defect)
The APRÈS outputs are **semantically approximate** even though well-formatted — e.g. "Gradient descent = Décomposition du gradient en deux dimensions" (not what GD is), "Sharpe ratio = ... seuil de rebond" (garbled). This is **exactly** the body's §8 honesty caveat ("le LoRA apprend le format et le style concis, pas la justesse encyclopédique profonde... contenu reste approximatif"). The committed wrong-but-formatted text is *evidence* of that limitation, surfaced honestly rather than hidden. Good discipline — the model's failure mode (format ✓, semantics ✗) is visible to the student.

## VRAM probe — real, measured (code#1 ec=2 / code#6 ec=7)
`torch.cuda.max_memory_allocated()/1e9`:
- **VRAM apres chargement : 0.47 Go / 8 Go** — matches body "0,47 Go".
- **VRAM pic pendant l'entrainement : 2.72 Go / 8 Go** — body said "2,68 Go"; **0.04 Go delta** (2.72 vs 2.68), immaterial rounding/non-determinism between body snapshot and committed run. Viable on 8 GB confirmed firsthand.

## 3 C.1 exercises — honest stubs
3 `# TODO etudiant` + 3 "à compléter" markers (r=2 vs r=32 / étendre dataset / save-reload adapter). `grep NotImplementedError|assert False` in source = **0**. Stubs honest, not fake.

## Verified
- **0 source churn**: NEW notebook (base = no file). README/catalogue untouched (catalogue auto-registers via cron <24h per body §E, separate readme grain honestly deferred).
- **Leak CLEAN**: 0 CJK (HARD RULE), 0 `sk-`/`ghp_`/`AKIA` secrets, 0 `C:\Users`, 0 `/home/<user>`, 0 broad `[A-Z]:\\` win-path. Model loads from public HuggingFace, no token.
- **FR-first**: prose pédagogique en français (spot-checked section headings + comparison output).

## △ Disclosure
No kernel/torch local → static verification (blob read, ec/outputs traced, SOTA markers confirmed, delta + VRAM values read from committed outputs, leak scan), NOT a re-execution of the LoRA training. Decisive here because the change is a NEW additive notebook (0 base touched → no regression risk) and the load-bearing Prong-B delta is **in the committed outputs** (re-exec not required to verify the format-acquisition claim). Trust CI over review if rouge.

## △ Nits (non-blocking)
- VRAM pic 2.72 Go (committed) vs body "2,68 Go" — 0.04 Go discrepancy, immaterial, note only.
- The APRÈS definitions being semantically wrong is pedagogically honest (see above) but a sharp student could read "APRÈS=True" as "the LoRA *understands* the term" — the §8 caveat mitigates this; could be reinforced in-cell next to the comparison, but not blocking.
- `prepare_model_for_kbit_training` not called explicitly — PEFT ≥0.13 handles k-bit prep internally via `get_peft_model` on a 4-bit loaded model; not a gap, just noting it's the newer convention vs the older explicit call.

Same genai-notebook vein as PT-08 (#10245, same #1454 post-training epic) and the 10_LocalLlama/11_Quantization arc this completes. Clean NEW-notebook PEFT/LoRA pedagogy — the Prong-B discrimination is measured (not pitched), the honest limitation is surfaced in real outputs, and the LoRA training is genuinely invoked end-to-end.
